### PR TITLE
Optimize snapshotting

### PR DESCRIPTION
**Pull request by beberlei**
Not sure tests ran on this, please verify implementation, related to #8

### CODE REVIEW

File: hdr.go [hunk 1]
```diff
@@ -292,21 +292,33 @@ func (h *Histogram) Equals(other *Histogram) bool {
...
 // Import to construct a new Histogram with the same state.
 func (h *Histogram) Export() *Snapshot {
+	var counts []int64
+	found := int64(0)
+
+	for i := int32(0); i < h.countsLen; i++ {
+		if found >= h.totalCount {
+			break
+		}
+		counts[i] = h.counts[i]
```
Comment (tylertreat) [inline hdr.go]: There seem to be at least a couple things wrong here. `counts` is never allocated and `found` is not being incremented.

File: hdr.go [hunk 2]
```diff
@@ -292,21 +292,34 @@ func (h *Histogram) Equals(other *Histogram) bool {
 // Export returns a snapshot view of the Histogram. This can be later passed to
 // Import to construct a new Histogram with the same state.
 func (h *Histogram) Export() *Snapshot {
+	counts := make([]int64, 0, len(h.counts))
```
Comment (beberlei) [inline hdr.go]: @tylertreat Is this here correct to increase the size of the counts as need fit and not make it as big as `h.counts` is already?

I will add a test to be sure I guess :-)

Comment (beberlei) [inline hdr.go]: It doesnt seem to work correctly.

**Comment by beberlei** [comment]
@tylertreat I refactored the code again, now it works, however it needs two loops over the data. I do think this is acceptable though. One option could be to have to export mechanisms, one optimized and the old one.

**Comment by ahothan** [comment]
Suggest to use the Encode and Decode feature that was just committed.
Will close this issue unless there is any objection.
Thanks

**Pull request closed without merging**